Integrate nrfx 3.0.0

awojasinski opened the pull request.

Commit:
drivers: sensor: qdec_nrfx: Add multi-instance support for QDEC SHIM

Reworked QDEC SHIM to suppor multi-instance peripheral. Patch includes
Kconfig alignment for proper instance handling.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: clock_control_nrf: Align LF clock sources symbols to new nrfx

New nrfx release brings change of Low frequency sources symbols
in nrf_clock hal to uppercase. This commit aligns all occurrences.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
counter: counter_nrfx_rtc: Align to renamed nrfx symbols

New nrfx release brings renamed macros and symbols in nrf_rtc.
This commit alligns RTC counter SHIM to it.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
samples: nrf: nrfx_prs: Align API to new nrfx release

New nrfx release brings API changes. This commit aligns sample to them.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
tests: boards: nrf52_bsim: egu: Replace obsolete timer frequency API

New nrfx release removes depricated `nrf_timer_frequency_set()`
function. Instead of it `nrf_timer_prescaler_set` is used with
`NRF_TIMER_PRESCALER_CALCULATE()` and `NRF_TIMER_BASE_FREQUENCY_GET()`
macros.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: watchdog: wdt_nrfx: Align behaviour parameter

New nrfx release replaces `nrf_wdt_behaviour_t` type
to `nrf_wdt_behaviour_mask_t` enumerator. As a result function
setting behaviour accepts `uint32_t` bitmask of behaviour.

This commit aligns symbols used in SHIM to the ones introduced
with the nrfx release.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: pwm: Align SHIM to renamed fields in driver instance struct

New nrfx release aligns PWM fields in driver instance structure
to common nrfx naming convention.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: spi: spi_nrfx_spim: Remove nrf_frequency_t handling

With new nrfx release SPIM driver stores frequency as a `uint32_t`
type representing frequency in `Hz` in its configuration structure.
Additionally `NRFX_SPIM_PIN_NOT_USED` has been removed,
`NRF_SPIM_PIN_NOT_CONNECTED` symbol is used instead.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: i2s: i2s_nrfx: Align SHIM to nrfx_i2s multi-instance API

New nrfx release extended nrfx_i2s API and requires to specify
pointer to driver instance structure. This commit aligns SHIM
to reworked nrfx driver.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
modules: hal_nordic: nrfx_config: Update nrfx_config files to nrfx 3.0

Update config files to recently released with new nrfx 3.0

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
manifest: hal_nordic: Update hal_nordic revision

Pull in nrfx 3.0.0 release

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
drivers: timer: nrf_rtc_timer: Align to renamed nrfx symbols

New nrfx release introduces renamed preprocesor symbols and macros
in nrf_rtc.h

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
modules: hal_nordic: nrfx_glue: Add cache auxiliary macros

Add cache auxiliary macros needed nrfy layer in nrfx drivers SW stack.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
modules: hal_nordic: CMakeLists: Add source files for GPPI helper

nrfx 3.0.0 splits implementation of GPPI helper functions into
periph-specific source files.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
samples: boards: nrf: nrfx: Align nrfx_gpiote API

nrfx 3.0.0 introduces renamed API for GPIOTE nrfx driver

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

Commit:
manifest: Update trusted-firmware-m revision

Pull API changes in platform nordic_nrf. Align use of GPIO API
to nrfx 3.0.0 release.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

awojasinski-nordicsemi commented on modules/hal_nordic/nrfx/nrfx_glue.h:
```diff
@@ -251,6 +251,45 @@ void nrfx_busy_wait(uint32_t usec_to_wait);
+ * @param[in] p_buffer Pointer to the buffer.
+ * @param[in] size     Size of the buffer.
+ */
+#define NRFY_CACHE_WB(p_buffer, size) \
+    do {                              \
+        (void)p_buffer;               \
+        (void)size;                   \
+    } while (0)
```
I'm not sure if compiler won't complain about unused variables

anangl replied:
It's a macro, so these are not variables.

awojasinski-nordicsemi replied:
but it might be called in a function where `p_buffer` is used only in the macro. Wouldn't that cause a compilation error is such a case? I'll remove it and we'll see

awojasinski-nordicsemi commented on modules/hal_nordic/nrfx/nrfx_config_nrf52810.h:
```diff
@@ -11,1803 +11,1232 @@
+ */
 #ifndef NRFX_CLOCK_CONFIG_CT_ENABLED
-#define NRFX_CLOCK_CONFIG_CT_ENABLED 0
+#define NRFX_CLOCK_CONFIG_CT_ENABLED 1
```
Yes, the symbol is always set for this target. Even if not in config file it is set in `nrfx_clock.c`  https://github.com/nika-nordic/hal_nordic/blob/a782df4518fe351ba54797d04d5e43e7336e8506/nrfx/drivers/src/nrfx_clock.c#L115

anangl replied:
My point is that it was set to 0 so far:
https://github.com/zephyrproject-rtos/zephyr/blob/f12ffaea1a9105d0cff74ea50ecde0549c680740/modules/hal_nordic/nrfx/nrfx_config_nrf52810.h#L44

nordic-krch commented on modules/hal_nordic/nrfx/nrfx_config_nrf51.h:
```diff
@@ -11,1212 +11,810 @@
+ * Integer value. Minimum: 0 Maximum: 15
+ */
+#ifndef NRFX_GPIOTE_CONFIG_NUM_OF_EVT_HANDLERS
+#define NRFX_GPIOTE_CONFIG_NUM_OF_EVT_HANDLERS 2
```
It can stay on `1`. More than 1 is needed for lpuart but that's handled by lpuart configuration

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -260,15 +219,82 @@ static int qdec_nrfx_pm_action(const struct device *dev,
+			.sampleper = NRF_QDEC_SAMPLEPER_2048US,					  \
+			.skip_gpio_cfg = true,							  \
+			.skip_psel_cfg = true,							  \
+			.ledpre  = QDEC_HAS_PROP(idx, led_pre),					  \
```
```suggestion
			.ledpre  = QDEC_PROP(idx, led_pre),					  \
```

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -260,15 +219,82 @@ static int qdec_nrfx_pm_action(const struct device *dev,
+#define SENSOR_NRFX_QDEC_DEVICE(idx)								  \
+	NRF_DT_CHECK_NODE_HAS_PINCTRL_SLEEP(DT_DRV_INST(idx));					  \
+	BUILD_ASSERT(QDEC_PROP(idx, steps) > 0,							  \
+		     "Wrong QDEC" #idx "steps setting in dts. Only positive number valid");	  \
+	BUILD_ASSERT(QDEC_PROP(idx, steps) <= 2048,						  \
+		     "Wrong QDEC" #idx "steps setting in dts. Overflow possible");		  \
```
```suggestion
		     "Wrong QDEC"#idx" steps setting in dts. Only positive number valid");	  \
	BUILD_ASSERT(QDEC_PROP(idx, steps) <= 2048,						  \
		     "Wrong QDEC"#idx" steps setting in dts. Overflow possible");		  \
```

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -260,15 +219,82 @@ static int qdec_nrfx_pm_action(const struct device *dev,
+	}											  \
+	static struct qdec_nrfx_data qdec_##idx##_data;						  \
+	PINCTRL_DT_INST_DEFINE(QDEC(idx));								  \
+	static struct qdec_nrfx_config *qdec_##idx##_config = {					  \
```
```suggestion
	static struct qdec_nrfx_config qdec_##idx##_config = {					  \
```

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -260,15 +219,82 @@ static int qdec_nrfx_pm_action(const struct device *dev,
+#define QDEC_PROP(idx, prop)		DT_PROP(QDEC(idx), prop)
+#define QDEC_HAS_PROP(idx, prop)	DT_NODE_HAS_PROP(QDEC(idx), prop)
+
+#define SENSOR_NRFX_QDEC_DEVICE(idx)								  \
```
Could you use shorter lines in this macro definition?

anangl commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -136,8 +136,17 @@ config NRFX_PWM3
+
+config NRFX_QDEC0
+	bool "QDEC0 driver instance"
+	depends on $(dt_nodelabel_has_compat,,$(DT_COMPAT_NORDIC_NRF_QDEC))
```
```suggestion
	depends on $(dt_nodelabel_has_compat,qdec0,$(DT_COMPAT_NORDIC_NRF_QDEC))
```

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -178,6 +178,12 @@
 #define NRFX_QDEC_ENABLED 1
 #endif
+#ifdef CONFIG_NRFX_QDEC0
+#define NRFX_QDEC_ENABLED 1
```
#define NRFX_QDEC0_ENABLED 1

anangl commented on drivers/sensor/qdec_nrfx/qdec_nrfx.c:
```diff
@@ -163,79 +154,47 @@ static void qdec_nrfx_event_handler(nrfx_qdec_event_t event)
 
-	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
-		    nrfx_isr, nrfx_qdec_irq_handler, 0);
+	if (config->enable_pin != NRF_QDEC_LED_NOT_CONNECTED) {
```
```suggestion
	if (config->enable_pin != NRF_QDEC_PIN_NOT_CONNECTED) {
```

Commit:
samples: sensor: qdec: Add overlays for Nordic Semiconductor boards

Add overlay for nrf52840dk_nrf52840 and nrf5340dk_nrf5340_cpuapp

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

anangl commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -14,3 +14,18 @@ tests:
         - "Position = (.*) degrees"
       fixture: fixture_mech_encoder
+  sample.sensor.nrf_qdec_sensor:
+    tags: sensors
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp
+    integration_platforms:
+      nrf52840dk_nrf52840
+      nrf5340dk_nrf5340_cpuapp
+    timeout: 5
+    harness: console
+    harness_config:
+      type: multi_line
+      ordered: true
+      regex:
+        - "Quadrature decoder sensor test"
+        - "Position = (.*) degrees"
```
You can specify most of this (except for `platform_allow` and `integration_platform`) under `common:` to not have to repeat it.

anangl commented on modules/hal_nordic/nrfx/nrfx_config_nrf91.h:
```diff
@@ -13,15 +13,16 @@
 
 #define NRF_CLOCK        NRF_PERIPH(NRF_CLOCK)
 #define NRF_DPPIC        NRF_PERIPH(NRF_DPPIC)
+#define NRF_DCNF         NRF_PERIPH(NRF_DCNF)
```
There is no such peripheral in nRF91.

Commit:
drivers: serial: uart_nrfx_uarte: Align use of TIMER macro in ASYNC API

nrfx 3.0 introduces parameter `_frequency` in
`NRFX_TIMER_DEFAULT_CONFIG()` macro. This patch aligns use of it.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

hubertmis commented on drivers/i2s/i2s_nrfx.c:
```diff
@@ -27,6 +27,7 @@ struct i2s_nrfx_drv_data {
 	struct k_msgq tx_queue;
 	struct stream_cfg rx;
 	struct k_msgq rx_queue;
+	const nrfx_i2s_t *p_i2s;
```
What `p` stands for in this field name?

awojasinski-nordicsemi commented:
force-pushed to rerun CI

anangl committed:
modules: hal_nordic: nrfx: Handle properly I2S0 instance

Although existing nRF SoCs have only one I2S instance, the nrfx_i2s
driver has now multi-instance API and the related nrfx configuration
symbols need to be used appropriately.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

fabiobaltieri commented:
Hi! This broke something with PWM, filed https://github.com/zephyrproject-rtos/zephyr/issues/57960 about it.
